File: ios-native/Dimo/Email/Gmail/GmailMessageParser.swift
import Foundation

struct ParsedGmailMessage: Hashable, Sendable {
  var gmailMessageId: String
  var gmailThreadId: String
  var rfcMessageId: String?
  var senderName: String?
  var senderAddress: String
  var subject: String
  var snippet: String
  var internalDate: Date
  var normalizedBody: String
}

enum GmailMessageParser {
  /// Attachment IDs for text body parts that Gmail omitted from the message payload.
  /// Large `text/plain` / `text/html` parts often come back with only `attachmentId`.
  static func unresolvedBodyAttachmentIds(from message: GmailMessageResource) -> [String] {
    var ids: [String] = []
    collectBodyParts(from: message.payload) { _, body in
      if let attachmentId = body.attachmentId?.trimmingCharacters(in: .whitespacesAndNewlines),
         !attachmentId.isEmpty,
         (body.data ?? "").isEmpty {
        ids.append(attachmentId)
      }
    }
    return ids.uniquedPreservingOrder()
  }

  static func parse(
    _ message: GmailMessageResource,
    resolvedBodies: [String: Data] = [:]
  ) throws -> ParsedGmailMessage {
    let headers = message.payload?.headers ?? []
    let sender = parseSender(header(headers, named: "From") ?? "")
    let subject = decodeHeaderValue(header(headers, named: "Subject") ?? "")
    let rfcMessageId = header(headers, named: "Message-ID")
      ?? header(headers, named: "Message-Id")
    guard let milliseconds = Int64(message.internalDate) else {
      throw GmailMessageParserError.invalidInternalDate
    }

    let body = selectedBody(from: message.payload, resolvedBodies: resolvedBodies)
      ?? message.snippet
      ?? ""
    return ParsedGmailMessage(
      gmailMessageId: message.id,
      gmailThreadId: message.threadId,
      rfcMessageId: rfcMessageId?.trimmingCharacters(in: .whitespacesAndNewlines),
      senderName: sender.name,
      senderAddress: sender.address,
      subject: subject,
      snippet: normalizePlainText(message.snippet ?? ""),
      internalDate: Date(timeIntervalSince1970: TimeInterval(milliseconds) / 1_000),
      normalizedBody: extractReadableText(body)
    )
  }

  /// Converts HTML/markup into readable plain text for storage and analysis.
  static func extractReadableText(_ raw: String) -> String {
    let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return "" }
    if looksLikeHTML(trimmed) {
      return sanitizeHTML(trimmed)
    }
    return normalizePlainText(trimmed)
  }

  static func sanitizeHTML(_ html: String) -> String {
    var value = html
    // Drop comments and non-content blocks, including unclosed style/script tails.
    value = value.replacingOccurrences(
      of: #"(?is)<!--.*?-->|<(script|style|head|svg|template|noscript)\b[^>]*>.*?(</\1\s*>|$)"#,
      with: " ",
      options: .regularExpression
    )
    value = value.replacingOccurrences(
      of: #"(?i)<\s*br\s*/?\s*>|</\s*(p|div|li|tr|h[1-6]|section|article|table|blockquote|pre)\s*>"#,
      with: "\n",
      options: .regularExpression
    )
    value = value.replacingOccurrences(
      of: #"(?i)</\s*td\s*>|</\s*th\s*>"#,
      with: "\t",
      options: .regularExpression
    )
    value = value.replacingOccurrences(of: #"(?is)<[^>]+>"#, with: " ", options: .regularExpression)
    value = decodeHTMLEntities(value)
    value = stripCSSLikeNoise(value)
    return normalizePlainText(value)
  }

  static func normalizePlainText(_ text: String) -> String {
    var value = text.precomposedStringWithCanonicalMapping
      .replacingOccurrences(of: "\u{00A0}", with: " ")
      .replacingOccurrences(of: "\r\n", with: "\n")
      .replacingOccurrences(of: "\r", with: "\n")
    value = String(value.unicodeScalars.filter { scalar in
      !CharacterSet.controlCharacters.contains(scalar)
        || scalar.value == 10 || scalar.value == 9
    })
    value = value.replacingOccurrences(of: #"[\t ]+"#, with: " ", options: .regularExpression)
    value = value.replacingOccurrences(of: #" *\n *"#, with: "\n", options: .regularExpression)
    value = value.replacingOccurrences(of: #"\n{3,}"#, with: "\n\n", options: .regularExpression)
    return value.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  static func decodeBase64URLData(_ encoded: String) -> Data? {
    var base64 = encoded.replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "_", with: "/")
    let padding = (4 - base64.count % 4) % 4
    base64 += String(repeating: "=", count: padding)
    return Data(base64Encoded: base64, options: .ignoreUnknownCharacters)
  }

  private static func selectedBody(
    from root: GmailMessagePart?,
    resolvedBodies: [String: Data]
  ) -> String? {
    var plainParts: [String] = []
    var htmlParts: [String] = []

    collectBodyParts(from: root) { mimeType, body in
      guard let text = decodeBodyText(body, resolvedBodies: resolvedBodies) else { return }
      if mimeType == "text/plain" {
        plainParts.append(text)
      } else if mimeType == "text/html" {
        htmlParts.append(text)
      }
    }

    let plainRaw = plainParts.joined(separator: "\n\n")
    let htmlRaw = htmlParts.joined(separator: "\n\n")
    return preferredReadableBody(plainRaw: plainRaw, htmlRaw: htmlRaw)
  }

  private static func preferredReadableBody(plainRaw: String, htmlRaw: String) -> String? {
    let plainText = plainRaw.isEmpty ? nil : extractReadableText(plainRaw).nilIfEmpty
    let htmlText = htmlRaw.isEmpty ? nil : sanitizeHTML(htmlRaw).nilIfEmpty
    switch (plainText, htmlText) {
    case (nil, nil):
      return nil
    case (let plain?, nil):
      return plain
    case (nil, let html?):
      return html
    case (let plain?, let html?):
      // Decide from the raw MIME parts: merchant "plain" is often HTML source
      // or a short "view in browser" stub.
      if looksLikeHTML(plainRaw) || looksLikeMarkupHeavy(plainRaw) || isHTMLViewerStub(plainRaw) {
        return html
      }
      if plain.count < 48, html.count > max(96, plain.count * 3) {
        return html
      }
      return plain
    }
  }

  private static func looksLikeHTML(_ text: String) -> Bool {
    let sample = String(text.prefix(8_000)).lowercased()
    if sample.contains("<html")
      || sample.contains("<!doctype")
      || sample.contains("<body")
      || sample.contains("<div")
      || sample.contains("<table")
      || sample.contains("<span")
      || sample.contains("<br")
      || sample.contains("<p ")
      || sample.contains("<p>") {
      return true
    }
    guard let regex = try? NSRegularExpression(pattern: #"<[a-z][\w:-]*\b"#) else {
      return false
    }
    let range = NSRange(sample.startIndex..., in: sample)
    return regex.numberOfMatches(in: sample, range: range) >= 4
  }

  private static func looksLikeMarkupHeavy(_ text: String) -> Bool {
    let sample = String(text.prefix(4_000))
    let angleCount = sample.filter { $0 == "<" || $0 == ">" }.count
    let braceCount = sample.filter { $0 == "{" || $0 == "}" }.count
    guard sample.count >= 40 else { return false }
    return Double(angleCount) / Double(sample.count) > 0.04
      || Double(braceCount) / Double(sample.count) > 0.05
  }

  private static func isHTMLViewerStub(_ text: String) -> Bool {
    let lower = text.lowercased()
    let hints = [
      "view this email in your browser",
      "view in browser",
      "enable html",
      "html is required",
      "doesn't support html",
      "does not support html",
      "multipart/alternative",
    ]
    return text.count < 280 && hints.contains { lower.contains($0) }
  }

  private static func stripCSSLikeNoise(_ text: String) -> String {
    text.split(separator: "\n", omittingEmptySubsequences: false)
      .filter { line in
        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return true }
        if trimmed == "{" || trimmed == "}" { return false }
        if trimmed.hasSuffix("{") { return false }
        if trimmed.contains("{"), trimmed.contains("}") { return false }
        // Compact CSS declarations: font-size:14px;color:#333;
        if trimmed.contains(":"),
           trimmed.contains(";"),
           trimmed.range(of: #"[\w\-]+\s*:\s*[^;{]+;"#, options: .regularExpression) != nil,
           trimmed.split(whereSeparator: \.isWhitespace).count <= 4 {
          return false
        }
        return true
      }
      .joined(separator: "\n")
  }

  private static func collectBodyParts(
    from root: GmailMessagePart?,
    visit: (_ mimeType: String, _ body: GmailMessageBody) -> Void
  ) {
    guard let root else { return }

    func walk(_ part: GmailMessagePart) {
      // Skip real file attachments (non-empty filename). Large inline text/html
      // bodies often use attachmentId with an empty filename and must be kept.
      let filename = (part.filename ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
      if !filename.isEmpty {
        return
      }

      let mimeType = (part.mimeType ?? "").lowercased()
      if let body = part.body {
        if mimeType.hasPrefix("text/plain") {
          visit("text/plain", body)
        } else if mimeType.hasPrefix("text/html") {
          visit("text/html", body)
        }
      }
      for child in part.parts ?? [] { walk(child) }
    }

    walk(root)
  }

  private static func decodeBodyText(
    _ body: GmailMessageBody,
    resolvedBodies: [String: Data]
  ) -> String? {
    if let encoded = body.data, !encoded.isEmpty,
       let data = decodeBase64URLData(encoded),
       let text = decodeText(data) {
      return text
    }
    if let attachmentId = body.attachmentId?.trimmingCharacters(in: .whitespacesAndNewlines),
       !attachmentId.isEmpty,
       let data = resolvedBodies[attachmentId],
       let text = decodeText(data) {
      return text
    }
    return nil
  }

  private static func header(_ headers: [GmailMessageHeader], named name: String) -> String? {
    headers.first { $0.name.caseInsensitiveCompare(name) == .orderedSame }?.value
  }

  private static func decodeText(_ data: Data) -> String? {
    String(data: data, encoding: .utf8)
      ?? String(data: data, encoding: .isoLatin1)
  }

  private static func parseSender(_ rawValue: String) -> (name: String?, address: String) {
    let decoded = decodeHeaderValue(rawValue).trimmingCharacters(in: .whitespacesAndNewlines)
    if let start = decoded.lastIndex(of: "<"),
       let end = decoded[start...].firstIndex(of: ">") {
      let address = decoded[decoded.index(after: start)..<end]
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .lowercased()
      let rawName = decoded[..<start].trimmingCharacters(in: .whitespacesAndNewlines)
        .trimmingCharacters(in: CharacterSet(charactersIn: "\""))
      return (rawName.isEmpty ? nil : rawName, address)
    }
    let addressPattern = #"[A-Z0-9._%+\-]+@[A-Z0-9.\-]+\.[A-Z]{2,}"#
    if let range = decoded.range(of: addressPattern, options: [.regularExpression, .caseInsensitive]) {
      return (nil, String(decoded[range]).lowercased())
    }
    return (nil, decoded.lowercased())
  }

  /// Decodes the common single-part RFC 2047 forms used by merchant senders.
  private static func decodeHeaderValue(_ value: String) -> String {
    let pattern = #"=\?([^?]+)\?([bBqQ])\?([^?]+)\?="#
    guard let regex = try? NSRegularExpression(pattern: pattern) else { return value }
    let range = NSRange(value.startIndex..., in: value)
    var result = value
    for match in regex.matches(in: value, range: range).reversed() {
      guard let whole = Range(match.range(at: 0), in: value),
            let encodingRange = Range(match.range(at: 2), in: value),
            let payloadRange = Range(match.range(at: 3), in: value) else { continue }
      let encoding = value[encodingRange].lowercased()
      let payload = String(value[payloadRange])
      let data: Data?
      if encoding == "b" {
        data = Data(base64Encoded: payload)
      } else {
        let q = payload.replacingOccurrences(of: "_", with: " ")
        data = decodeQuotedPrintable(q)
      }
      guard let data, let decoded = decodeText(data) else { continue }
      result.replaceSubrange(whole, with: decoded)
    }
    return result
  }

  private static func decodeQuotedPrintable(_ value: String) -> Data? {
    var bytes: [UInt8] = []
    let input = Array(value.utf8)
    var index = 0
    while index < input.count {
      if input[index] == 61, index + 2 < input.count,
         let high = hex(input[index + 1]), let low = hex(input[index + 2]) {
        bytes.append(high * 16 + low)
        index += 3
      } else {
        bytes.append(input[index])
        index += 1
      }
    }
    return Data(bytes)
  }

  private static func hex(_ value: UInt8) -> UInt8? {
    switch value {
    case 48...57: return value - 48
    case 65...70: return value - 55
    case 97...102: return value - 87
    default: return nil
    }
  }

  private static func decodeHTMLEntities(_ value: String) -> String {
    let named: [String: String] = [
      "&amp;": "&", "&lt;": "<", "&gt;": ">", "&quot;": "\"",
      "&#39;": "'", "&apos;": "'", "&nbsp;": " ", "&ndash;": "–", "&mdash;": "—",
    ]
    var result = value
    for (entity, replacement) in named {
      result = result.replacingOccurrences(of: entity, with: replacement, options: .caseInsensitive)
    }
    guard let regex = try? NSRegularExpression(pattern: #"&#(x[0-9a-fA-F]+|[0-9]+);"#) else {
      return result
    }
    for match in regex.matches(in: result, range: NSRange(result.startIndex..., in: result)).reversed() {
      guard let whole = Range(match.range(at: 0), in: result),
            let numberRange = Range(match.range(at: 1), in: result) else { continue }
      let raw = String(result[numberRange])
      let number = raw.lowercased().hasPrefix("x")
        ? UInt32(raw.dropFirst(), radix: 16)
        : UInt32(raw, radix: 10)
      if let number, let scalar = UnicodeScalar(number) {
        result.replaceSubrange(whole, with: String(scalar))
      }
    }
    return result
  }
}

enum GmailMessageParserError: LocalizedError {
  case invalidInternalDate

  var errorDescription: String? { "The Gmail message date is invalid." }
}

private extension Array where Element: Hashable {
  func uniquedPreservingOrder() -> [Element] {
    var seen = Set<Element>()
    return filter { seen.insert($0).inserted }
  }
}

private extension String {
  var nilIfEmpty: String? {
    trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self
  }
}

File: android-native/app/src/main/java/app/dimo/android/email/gmail/GmailMessageParser.kt
package app.dimo.android.email.gmail

import android.util.Base64
import java.text.Normalizer

/**
 * Port of `ios-native/Dimo/Email/Gmail/GmailMessageParser.swift`.
 *
 * Turns a Gmail message resource into the normalized plain text the analyzer
 * sees. Merchant mail is adversarially messy — HTML masquerading as `text/plain`,
 * "view in browser" stubs, CSS dumped into the body — so body selection is
 * evidence-based rather than trusting the declared MIME type.
 */
data class ParsedGmailMessage(
  val gmailMessageId: String,
  val gmailThreadId: String,
  val rfcMessageId: String?,
  val senderName: String?,
  val senderAddress: String,
  val subject: String,
  val snippet: String,
  val internalDate: Long,
  val normalizedBody: String,
)

object GmailMessageParser {
  /**
   * Attachment IDs for text body parts Gmail omitted from the message payload.
   * Large `text/plain` / `text/html` parts often come back with only `attachmentId`.
   */
  fun unresolvedBodyAttachmentIds(message: GmailMessageResource): List<String> {
    val ids = mutableListOf<String>()
    collectBodyParts(message.payload) { _, body ->
      val attachmentId = body.attachmentId?.trim()
      if (!attachmentId.isNullOrEmpty() && body.data.isNullOrEmpty()) ids.add(attachmentId)
    }
    return ids.distinct()
  }

  fun parse(
    message: GmailMessageResource,
    resolvedBodies: Map<String, ByteArray> = emptyMap(),
  ): ParsedGmailMessage {
    val headers = message.payload?.headers.orEmpty()
    val sender = parseSender(header(headers, "From").orEmpty())
    val subject = decodeHeaderValue(header(headers, "Subject").orEmpty())
    val rfcMessageId = header(headers, "Message-ID") ?: header(headers, "Message-Id")
    val milliseconds = message.internalDate.toLongOrNull()
      ?: throw GmailMessageParserException("The Gmail message date is invalid.")

    val body = selectedBody(message.payload, resolvedBodies) ?: message.snippet.orEmpty()
    return ParsedGmailMessage(
      gmailMessageId = message.id,
      gmailThreadId = message.threadId,
      rfcMessageId = rfcMessageId?.trim(),
      senderName = sender.first,
      senderAddress = sender.second,
      subject = subject,
      snippet = normalizePlainText(message.snippet.orEmpty()),
      internalDate = milliseconds,
      normalizedBody = extractReadableText(body),
    )
  }

  /** Converts HTML/markup into readable plain text for storage and analysis. */
  fun extractReadableText(raw: String): String {
    val trimmed = raw.trim()
    if (trimmed.isEmpty()) return ""
    return if (looksLikeHtml(trimmed)) sanitizeHtml(trimmed) else normalizePlainText(trimmed)
  }

  fun sanitizeHtml(html: String): String {
    var value = html
    // Drop comments and non-content blocks, including unclosed style/script tails.
    value = NON_CONTENT_BLOCKS.replace(value, " ")
    value = LINE_BREAK_TAGS.replace(value, "\n")
    value = CELL_END_TAGS.replace(value, "\t")
    value = ANY_TAG.replace(value, " ")
    value = decodeHtmlEntities(value)
    value = stripCssLikeNoise(value)
    return normalizePlainText(value)
  }

  fun normalizePlainText(text: String): String {
    var value = Normalizer.normalize(text, Normalizer.Form.NFC)
      .replace(" ", " ")
      .replace("\r\n", "\n")
      .replace("\r", "\n")
    // Keep newline and tab; drop every other control character.
    value = value.filter { it == '\n' || it == '\t' || !it.isISOControl() }
    value = HORIZONTAL_SPACE.replace(value, " ")
    value = PADDED_NEWLINE.replace(value, "\n")
    value = EXCESS_NEWLINES.replace(value, "\n\n")
    return value.trim()
  }

  fun decodeBase64Url(encoded: String): ByteArray? = runCatching {
    Base64.decode(encoded, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
  }.getOrNull()

  // MARK: - Body selection

  private fun selectedBody(
    root: GmailMessagePart?,
    resolvedBodies: Map<String, ByteArray>,
  ): String? {
    val plainParts = mutableListOf<String>()
    val htmlParts = mutableListOf<String>()
    collectBodyParts(root) { mimeType, body ->
      val text = decodeBodyText(body, resolvedBodies) ?: return@collectBodyParts
      when (mimeType) {
        "text/plain" -> plainParts.add(text)
        "text/html" -> htmlParts.add(text)
      }
    }
    return preferredReadableBody(
      plainRaw = plainParts.joinToString("\n\n"),
      htmlRaw = htmlParts.joinToString("\n\n"),
    )
  }

  private fun preferredReadableBody(plainRaw: String, htmlRaw: String): String? {
    val plainText = plainRaw.takeIf { it.isNotEmpty() }?.let { extractReadableText(it).nilIfBlank() }
    val htmlText = htmlRaw.takeIf { it.isNotEmpty() }?.let { sanitizeHtml(it).nilIfBlank() }
    if (plainText == null) return htmlText
    if (htmlText == null) return plainText
    // Decide from the raw MIME parts: merchant "plain" is often HTML source or a
    // short "view in browser" stub.
    if (looksLikeHtml(plainRaw) || looksLikeMarkupHeavy(plainRaw) || isHtmlViewerStub(plainRaw)) {
      return htmlText
    }
    if (plainText.length < 48 && htmlText.length > maxOf(96, plainText.length * 3)) {
      return htmlText
    }
    return plainText
  }

  private fun looksLikeHtml(text: String): Boolean {
    val sample = text.take(8_000).lowercase()
    val markers = listOf(
      "<html", "<!doctype", "<body", "<div", "<table", "<span", "<br", "<p ", "<p>",
    )
    if (markers.any { sample.contains(it) }) return true
    return OPEN_TAG.findAll(sample).take(4).count() >= 4
  }

  private fun looksLikeMarkupHeavy(text: String): Boolean {
    val sample = text.take(4_000)
    if (sample.length < 40) return false
    val angles = sample.count { it == '<' || it == '>' }
    val braces = sample.count { it == '{' || it == '}' }
    return angles.toDouble() / sample.length > 0.04 || braces.toDouble() / sample.length > 0.05
  }

  private fun isHtmlViewerStub(text: String): Boolean {
    if (text.length >= 280) return false
    val lower = text.lowercase()
    return VIEWER_STUB_HINTS.any { lower.contains(it) }
  }

  private fun stripCssLikeNoise(text: String): String =
    text.split("\n").filter { line ->
      val trimmed = line.trim()
      when {
        trimmed.isEmpty() -> true
        trimmed == "{" || trimmed == "}" -> false
        trimmed.endsWith("{") -> false
        trimmed.contains("{") && trimmed.contains("}") -> false
        // Compact CSS declarations: font-size:14px;color:#333;
        trimmed.contains(":") && trimmed.contains(";") &&
          CSS_DECLARATION.containsMatchIn(trimmed) &&
          trimmed.split(WHITESPACE).size <= 4 -> false
        else -> true
      }
    }.joinToString("\n")

  private fun collectBodyParts(
    root: GmailMessagePart?,
    visit: (mimeType: String, body: GmailMessageBody) -> Unit,
  ) {
    if (root == null) return

    fun walk(part: GmailMessagePart) {
      // Skip real file attachments (non-empty filename). Large inline text/html
      // bodies often use attachmentId with an empty filename and must be kept.
      if (!part.filename.orEmpty().trim().isEmpty()) return
      val mimeType = part.mimeType.orEmpty().lowercase()
      part.body?.let { body ->
        when {
          mimeType.startsWith("text/plain") -> visit("text/plain", body)
          mimeType.startsWith("text/html") -> visit("text/html", body)
        }
      }
      part.parts.orEmpty().forEach(::walk)
    }

    walk(root)
  }

  private fun decodeBodyText(
    body: GmailMessageBody,
    resolvedBodies: Map<String, ByteArray>,
  ): String? {
    body.data?.takeIf { it.isNotEmpty() }?.let { encoded ->
      decodeBase64Url(encoded)?.let { data -> decodeText(data)?.let { return it } }
    }
    body.attachmentId?.trim()?.takeIf { it.isNotEmpty() }?.let { attachmentId ->
      resolvedBodies[attachmentId]?.let { data -> decodeText(data)?.let { return it } }
    }
    return null
  }

  private fun header(headers: List<GmailMessageHeader>, name: String): String? =
    headers.firstOrNull { it.name.equals(name, ignoreCase = true) }?.value

  /** UTF-8 with a Latin-1 fallback, matching the Swift decoder chain. */
  private fun decodeText(data: ByteArray): String? =
    runCatching { data.toString(Charsets.UTF_8) }.getOrNull()
      ?: runCatching { data.toString(Charsets.ISO_8859_1) }.getOrNull()

  private fun parseSender(rawValue: String): Pair<String?, String> {
    val decoded = decodeHeaderValue(rawValue).trim()
    val start = decoded.lastIndexOf('<')
    val end = if (start >= 0) decoded.indexOf('>', start) else -1
    if (start >= 0 && end > start) {
      val address = decoded.substring(start + 1, end).trim().lowercase()
      val rawName = decoded.substring(0, start).trim().trim('"')
      return (rawName.takeIf { it.isNotEmpty() }) to address
    }
    ADDRESS_PATTERN.find(decoded)?.let { return null to it.value.lowercase() }
    return null to decoded.lowercase()
  }

  /** Decodes the common single-part RFC 2047 forms used by merchant senders. */
  private fun decodeHeaderValue(value: String): String {
    var result = value
    for (match in RFC2047.findAll(value).toList().asReversed()) {
      val encoding = match.groupValues[2].lowercase()
      val payload = match.groupValues[3]
      val data = if (encoding == "b") {
        runCatching { Base64.decode(payload, Base64.DEFAULT) }.getOrNull()
      } else {
        decodeQuotedPrintable(payload.replace('_', ' '))
      }
      val decoded = data?.let { decodeText(it) } ?: continue
      result = result.replaceRange(match.range, decoded)
    }
    return result
  }

  private fun decodeQuotedPrintable(value: String): ByteArray {
    val input = value.toByteArray(Charsets.UTF_8)
    val bytes = ArrayList<Byte>(input.size)
    var index = 0
    while (index < input.size) {
      val high = if (index + 2 < input.size) hex(input[index + 1]) else null
      val low = if (index + 2 < input.size) hex(input[index + 2]) else null
      if (input[index] == '='.code.toByte() && high != null && low != null) {
        bytes.add((high * 16 + low).toByte())
        index += 3
      } else {
        bytes.add(input[index])
        index += 1
      }
    }
    return bytes.toByteArray()
  }

  private fun hex(value: Byte): Int? = when (val code = value.toInt() and 0xFF) {
    in 48..57 -> code - 48
    in 65..70 -> code - 55
    in 97..102 -> code - 87
    else -> null
  }

  private fun decodeHtmlEntities(value: String): String {
    var result = value
    for ((entity, replacement) in NAMED_ENTITIES) {
      result = result.replace(entity, replacement, ignoreCase = true)
    }
    for (match in NUMERIC_ENTITY.findAll(result).toList().asReversed()) {
      val raw = match.groupValues[1]
      val number = if (raw.startsWith("x", ignoreCase = true)) {
        raw.drop(1).toIntOrNull(16)
      } else {
        raw.toIntOrNull()
      } ?: continue
      if (number !in 0..0x10FFFF) continue
      result = result.replaceRange(match.range, String(Character.toChars(number)))
    }
    return result
  }

  private fun String.nilIfBlank(): String? = takeIf { it.isNotBlank() }

  private val NON_CONTENT_BLOCKS = Regex(
    "<!--.*?-->|<(script|style|head|svg|template|noscript)\\b[^>]*>.*?(</\\1\\s*>|$)",
    setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL),
  )
  private val LINE_BREAK_TAGS = Regex(
    "<\\s*br\\s*/?\\s*>|</\\s*(p|div|li|tr|h[1-6]|section|article|table|blockquote|pre)\\s*>",
    RegexOption.IGNORE_CASE,
  )
  private val CELL_END_TAGS = Regex("</\\s*td\\s*>|</\\s*th\\s*>", RegexOption.IGNORE_CASE)
  private val ANY_TAG = Regex("<[^>]+>", RegexOption.DOT_MATCHES_ALL)
  private val OPEN_TAG = Regex("<[a-z][\\w:-]*\\b")
  private val HORIZONTAL_SPACE = Regex("[\\t ]+")
  private val PADDED_NEWLINE = Regex(" *\\n *")
  private val EXCESS_NEWLINES = Regex("\\n{3,}")
  private val CSS_DECLARATION = Regex("[\\w\\-]+\\s*:\\s*[^;{]+;")
  private val WHITESPACE = Regex("\\s+")
  private val ADDRESS_PATTERN =
    Regex("[A-Z0-9._%+\\-]+@[A-Z0-9.\\-]+\\.[A-Z]{2,}", RegexOption.IGNORE_CASE)
  private val RFC2047 = Regex("=\\?([^?]+)\\?([bBqQ])\\?([^?]+)\\?=")
  private val NUMERIC_ENTITY = Regex("&#(x[0-9a-fA-F]+|[0-9]+);")

  private val NAMED_ENTITIES = listOf(
    "&amp;" to "&", "&lt;" to "<", "&gt;" to ">", "&quot;" to "\"",
    "&#39;" to "'", "&apos;" to "'", "&nbsp;" to " ", "&ndash;" to "–", "&mdash;" to "—",
  )

  private val VIEWER_STUB_HINTS = listOf(
    "view this email in your browser",
    "view in browser",
    "enable html",
    "html is required",
    "doesn't support html",
    "does not support html",
    "multipart/alternative",
  )
}

class GmailMessageParserException(message: String) : Exception(message)
